### DISCUSSION TITLE
How to ensure all the guards and actions are implemented in setup?

### DISCUSSION THREAD

**Discussion by Vithop** [statelyai/xstate#4788] [Q&A]
Hey I have a use case where I have created a type to define all my `guards` and `actions` but I don't know what `state machine` is implemented at build time (I plan to have a API call which fetches this data). 

Currently the guards and actions in the setup function are configured to be optional so I can't statically determine if all my guards and actions have been implemented before calling `createMachine`

The following is an example of what I'm striving for.

```
enum NavigationGuardEnum {
  guardOne: "guardOne",
  guardTwo: "guardTwo"
}

enum NavigationActionEnum {
  actionOne: "actionOne",
  actionTwo: "actionTwo"
}

const myNavigationRouterMachine = setup({
  types: {} as {
    guards: { type: NavigationGuardEnum };
    actions: { type: NavigationActionEnum };
  },
  guards: { // <= I want this to error with undefined guards!!!
    guardOne: ({ context }) => return true
  },
  actions: { <= I want this to error with undefined actions!!!
    actionOne: ({ event }) => doSomethingCool(),
  },
})
```
I do know that when I call createMachine and reference a action or guard which is undefined it will throw a typescript error but in my case I won't know everything in the createMachine call until runtime. Therefore I'm trying to find a way to statically type this during setup.

**Comment by Andarist** [comment]
When using `setup` you should not be using `types.guards` and `types.actions`. Those types are already derived from the top-level keys and you can't constrain what has to be provided there using the `types` property.

**Reply by Vithop** [reply]
Is there a way to constrain the implemented actions and guards to the type? Like by using `provide`?

**Reply by Andarist** [reply]
`satisfies` would be perfect for this but it doesn't work like this today: https://github.com/microsoft/TypeScript/issues/57667

**Reply by Vithop** [reply]
Yeah it would be cool if `satisfies`  did the trick. But in my code base I'm having some trouble using satisfies with my tooling so I won't be able to use it any time soon either.
